kernel/src/dynamic_binary_storage.rs

```diff
@@ -0,0 +1,615 @@
+    ///   process, the function will return `Ok()` with the size of the region
+    ///   to store the process, and whether the process loader is waiting to set
+    ///   up. This usually happens when we have to write a post pad app. The
+    ///   client app is unable to write new app data until the process loader
+    ///   finishes writing the padding app. So if tihs flag is set, then the
+    ///   client app has to wait until the setup_done subscribe callback is
+    ///   received.
+    /// - `Err(ErrorCode)`: If there is nowhere to store the process a suitable
+    ///   `ErrorCode` will be returned.
+    fn setup(&self, app_length: usize) -> Result<(usize, bool), ErrorCode>;
+
+    /// Instruct the kernel to write data to the flash.
+    ///
+    /// `offset` is where to start writing within the region allocated
+    /// for the new process binary from the `setup()` call.
+    ///
+    /// The caller must write the first 8 bytes of the process with valid header
+    /// data. Writes must either be after the first 8 bytes or include the
+    /// entire first 8 bytes.
+    ///
+    /// Returns an error if the write is outside of the permitted region or is
+    /// writing an invalid header.
+    fn write_app_data(
+        &self,
+        buffer: SubSliceMut<'static, u8>,
+        offset: usize,
+    ) -> Result<(), ErrorCode>;
+
+    /// Sets a client for the SequentialDynamicBinaryStore Object
+    ///
+    /// When the client operation is done, it calls the `setup_done()`
+    /// and `write_app_data_done()` functions.
+    fn set_storage_client(&self, client: &'static dyn SequentialDynamicBinaryStoreClient);
+
+    /// Write a prepad app if required
+    fn write_prepad_app(&self);
+}
+
+/// The callback for dynamic binary flashing.
+pub trait SequentialDynamicBinaryStoreClient {
+    /// Any setup work is done and we are ready to write the process binary.
+    fn setup_done(&self);
+
+    /// The provided app binary buffer has been stored.
+    fn write_app_data_done(&self, buffer: &'static mut [u8], length: usize);
+}
+
+/// This interface supports loading processes at runtime.
+pub trait SequentialDynamicProcessLoad {
+    /// Call to request kernel to load a new process.
+    fn load(&self) -> Result<(), ErrorCode>;
+
+    /// Sets a client for the SequentialDynamicProcessLoading Object
+    ///
+    /// When the client operation is done, it calls the `load_done()`
+    /// function.
+    fn set_load_client(&self, client: &'static dyn SequentialDynamicProcessLoadClient);
+}
+
+/// The callback for dynamic binary flashing.
+pub trait SequentialDynamicProcessLoadClient {
+    /// The new app has been loaded.
+    fn load_done(&self);
+}
+
+/// Dynamic process loading machine.
+pub struct SequentialDynamicBinaryStorage<'a, C: Chip + 'static, D: ProcessStandardDebug + 'static>
+{
+    processes: MapCell<&'static mut [Option<&'static dyn process::Process>]>,
+    flash_driver: &'a dyn NonvolatileStorage<'a>,
```

Comment:
```suggestion
pub struct SequentialDynamicBinaryStorage<'a, C: Chip + 'static, D: ProcessStandardDebug + 'static>
{
    flash_driver: &'a dyn NonvolatileStorage<'a>,
```

We don't need this just for debugging.